sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -409,6 +505,340 @@ class GpuDynamicPartitionDataWriter(
+    GpuOutOfCoreSortIterator(iterator, sorter, cpuOrd, targetSize,
+      opTime, sortTime, outputBatch, outputRows,
+      peakDevMemory, spillCallback)
+  }
+
+  /**
+   * concurrent write the columnar batch
+   * @param cb the columnar batch
+   */
+  override def write(cb: ColumnarBatch): Unit = {
+    assert(isPartitioned)
+    assert(!isBucketed)
+
+    if (cb.numRows() == 0) {
+      // intend to write meta only
+      // TODO https://github.com/NVIDIA/spark-rapids/issues/6453
+      return
+    }
+
+    // 1. combine partition columns and `cb` columns into a column array
+    val columnsWithPartition = ArrayBuffer[ColumnVector]()
+    withResource(getPartitionColumns(cb)) { partitionColumnsTable =>
+      for (i <- 0 until partitionColumnsTable.getNumberOfColumns) {
+        // append partition column
+        columnsWithPartition += partitionColumnsTable.getColumn(i)
+      }
+    }
+    val cols = GpuColumnVector.extractBases(cb)
+    columnsWithPartition ++= cols
+
+    // 2. group by the partition columns
+    // get sub-groups for each partition and get unique keys for each partition
+    val groupsAndKeys = withResource(cb) { _ =>
+      // combine partition columns and `cb` columns
+      withResource(new Table(columnsWithPartition: _*)) { colsWithPartitionTbl =>
+        // [0, partition columns number - 1]
+        val partitionIndices = description.partitionColumns.indices
+
+        // group by partition columns
+        val op = colsWithPartitionTbl.groupBy(partitionIndices: _*)
+        // return groups and uniq keys table
+        // Each row in uniq keys table is corresponding to a group
+        op.contiguousSplitGroupsAndGenUniqKeys()
+      }
+    }
+
+    withResource(groupsAndKeys) { _ =>
+      // groups number should equal to uniq key number
+      assert(groupsAndKeys.getGroups.length == groupsAndKeys.getUniqKeyTable.getRowCount)
+
+      val (groups, keys) = (groupsAndKeys.getGroups, groupsAndKeys.getUniqKeyTable)
+
+      // 3. generate partition strings for all sub-groups in advance
+      val partDataTypes = description.partitionColumns.map(_.dataType).toArray
+      val dataTypes = GpuColumnVector.extractTypes(cb)
+      // generate partition string list for all groups
+      val partitionStrList = getPartitionStrList(keys, partDataTypes)
+      // key table is useless now
+      groupsAndKeys.closeUniqKeyTable()
+
+      // 4. cache each group according to each partitionStr
+      withResource(groups) { _ =>
+        var groupIndex = 0
+        while (groupIndex < groups.length) {
+          // get the partition string and group pair
+          val (partitionStr, group) = (partitionStrList(groupIndex), groups(groupIndex))
+          val groupTable = group.getTable
+          groupIndex += 1
+
+          // create writer if encounter a new partition and put into `concurrentWriters` map
+          if (!concurrentWriters.contains(partitionStr)) {
+            val w = newWriter(partitionStr, None, 0)
```

Comment:
Why are we not checking if we have too many writers here? Or preferably up above once we know all of the partitions?

Reply:
Moved to the place:
```
if (!concurrentWriters.contains(partitionStr)) {
   ...
   concurrentWriters.put(partitionStr, new WriterStatusWithCaches(ws))
   if (!fallBackToSortBased && concurrentWriters.size >= spec.maxWriters) {
              // Note: fall back will be trigger
              fallBackToSortBased = true
            }
```

But, note: when marked falling back to `single writer`, 
 
```
override def write(cb: ColumnarBatch): Unit
```

Reply:
So then the max number of writes that we have open is actually either the maximum number of writes needed to complete a batch?  I just want to be sure that we document this properly when we get to documentation.

Reply:
Updated.
If the number of the new partition in `cb` plus the number of the existing partition is greater than
`maxWriters` limit, will put the whole `cb` back to 'single writer`